llama_index/vector_stores/postgres.py

```diff
@@ -216,11 +263,132 @@ async def _aquery_with_score(
                         node_id=item.node_id,
                         text=item.text,
                         metadata=item.metadata_,
-                        similarity=(1 - distance),
+                        similarity=(1 - distance) if distance is not None else 0,
                     )
                     for item, distance in res.all()
                 ]
 
+    def _rerank(self, query, results):
+        try:
+            from sentence_transformers import CrossEncoder
+        except ImportError:
+            raise ImportError(
+                "Cannot import sentence-transformers package,",
+                "please `pip install sentence-transformers`",
+            )
+        import itertools
+
+        # Based on example code from: https://github.com/pgvector/pgvector-python/blob/master/examples/hybrid_search.py
+        # deduplicate
+        seen = set()
+        uniq_results = []
+        for result in itertools.chain(*results):
+            if result.node_id not in seen:
+                seen.add(result.node_id)
+                uniq_results.append(result)
+
+        # re-rank
+        encoder = CrossEncoder(self._hybrid_search_cross_encoder)
+        scores = encoder.predict([(query.query_str, item.text) for item in uniq_results])
+        return [v for _, v in sorted(zip(scores, uniq_results), reverse=True)]
+
+    def _build_sparse_query(
+            self,
+            query_str: Optional[str],
+            limit: int,
+            metadata_filters: Optional[MetadataFilters] = None,
+    ) -> Any:
+        import sqlalchemy
+        from sqlalchemy import select
+        from sqlalchemy.sql import func, text
+
+        if query_str is None:
+            raise ValueError(
+                "query_str must be specified for a sparse vector query."
+            )
+
+        stmt = select(  # type: ignore
+            self.table_class,
+            func.ts_rank(self.table_class.text_search_tsv, func.plainto_tsquery(query_str)).label('rank')
```

Comment:
This is currently using the `ts_rank()` function for now. I can also make this is a customizable argument, but currently the implementation is opinionated to reduce the number of flags/options presented to the user.